### PR TITLE
Bug fix/devsup 899

### PR DESCRIPTION
**Pull request by mchacki**
### Scope & Purpose

Fixes Devsup-899.
  in a very rare case a subquery, nested in another subquery, was not
  executed, which is fixed now. Technical details:
  If we have two subqueries: `Outer` and `Nested` the Outer will define
  the input for Nested. And Outer has the pattern: 1 input, subqueryDone, 1
  input, subqueryDone [...] and our internal batching did cut a batch like
  this: [<...>, input (A)] | [subqueryDone, input (B), subqueryDone, <...>]
  than Nested on input (B) was not executed. As soon as we have more than 1
  input per Outer, or a different cutting position, all was good.

- [x] :hankey: Bugfix (requires CHANGELOG entry)
- [ ] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification
- [ ] :book: CHANGELOG entry made

#### Backports:

- [x] Backport for 3.9: https://github.com/arangodb/arangodb/pull/15039
- [x] Backport for 3.8: https://github.com/arangodb/arangodb/pull/15040
- [x] Backport for 3.7: https://github.com/arangodb/arangodb/pull/15041

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket number: Devsup-899
- [ ] Design document: 

### Testing & Verification

*(Please pick either of the following options)*

- [ ] This change is a trivial rework / code cleanup without any test coverage.
- [x] The behavior in this PR was *manually tested*
- [ ] This change is already covered by existing tests, such as *(please describe tests)*.
- [x] This PR adds tests that were used to verify all changes:
  - [x] Added new C++ **Unit tests**
  - [ ] Added new **integration tests** (e.g. in shell_server / shell_server_aql)
  - [ ] Added new **resilience tests** (only if the feature is impacted by failovers)
- [ ] There are tests in an external testing repository:
- [ ] I ensured this code runs with ASan / TSan or other static verification tools

Link to Jenkins PR run:

### Documentation

> All new features should be accompanied by corresponding documentation. 
> Bugs and features should furthermore be documented in the CHANGELOG so that
> support, end users, and other developers have a concise overview. 

- [ ] Added entry to *Release Notes* 
- [ ] Added a new section in the *Manual* 
- [ ] Added a new section in the *HTTP API* 
- [ ] Added *Swagger examples* for the HTTP API  
- [ ] Updated license information in *LICENSES-OTHER-COMPONENTS.md* for 3rd party libraries

### External contributors / CLA Note 

Please note that for legal reasons we require you to sign the [Contributor Agreement](https://www.arangodb.com/documents/cla.pdf)
before we can accept your pull requests.

### CODE REVIEW

File: tests/Aql/FixedOutputExecutionBlockMock.h [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+#include "Aql/RegisterInfos.h"
+#include "Aql/SharedAqlItemBlockPtr.h"
+
+#include <deque>
+
+namespace arangodb {
+
+namespace tests {
+namespace aql {
+
```
Comment (markuspf) [inline tests/Aql/FixedOutputExecutionBlockMock.h:35]: Maybe say what the purpose of this mock is (even if it is nigh-obvious I guess)

File: tests/Aql/FixedOutputExecutionBlockMock.cpp [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+#include "Aql/AqlCallStack.h"
+#include "Basics/voc-errors.h"
+
+using namespace arangodb;
+using namespace arangodb::aql;
+using namespace arangodb::tests::aql;
+
+namespace {
+// NOTE copy pasted from Waiting ExecutionBlock mock
```
Comment (markuspf) [inline tests/Aql/FixedOutputExecutionBlockMock.cpp:33]: If it's copy-and-pasted, could it live somewhere to be used in both cases?

**Review by markuspf — APPROVED** [review]
The fix looks good to me, though it's a bit embarassing that this has been here for such a long time (and I was likely involved in the making of it.).

Few comments about code style/comments.

File: tests/Aql/FixedOutputExecutionBlockMock.cpp (hunk 2 above)
Comment (mchacki) [reply tests/Aql/FixedOutputExecutionBlockMock.cpp:33]: I made them slightly different now.

File: tests/Aql/FixedOutputExecutionBlockMock.cpp [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+      regs = b->numRegisters();
+
+      break;
+    }
+  }
+  // if non found sorry blind guess the number of registers here.
+  // This can happen if you insert the data later into this Mock.
+  // If you do so this register planning is of
+  // for the rime being no test is showing this behavior.
+  // Consider adding data first if the test fails
```
[the remark this answers is not in the document]
Comment (mchacki) [reply tests/Aql/FixedOutputExecutionBlockMock.cpp:52]: Removed and replaced with a better description.

**Review by markuspf — APPROVED** [review]

**Pull request merged**